crates/lib/src/db/column_ordering.rs

```diff
@@ -0,0 +1,129 @@
+use std::cmp::Ordering;
+
+use spacetimedb_sats::{AlgebraicType, BuiltinType, Typespace};
+
+/// Canonical ordering for fields of a Table.
+///
+/// Fields are ordered by their `SpacetimeDB` alignment, then alphabetically by name.
+///
+/// Does not currently apply to all AlgebraicTypes.
+pub fn canonical_column_ordering(
+    typespace: &Typespace,
+    column_1: (&str, &AlgebraicType),
+    column_2: (&str, &AlgebraicType),
+) -> Ordering {
+    alignment(column_1.1, typespace)
+        .cmp(&alignment(column_2.1, typespace))
+        .reverse() // big goes before small
+        .then_with(|| column_1.0.cmp(column_2.0))
+}
+
+/// Determine the SpacetimeDB alignment of a value of this algebraic type.
+///
+/// Note: these alignment values are only used to choose a canonical serialization ordering.
+/// They are not used to actually align data in memory.
+/// Really, we could use any ordering here, as long as it is consistent.
+///
+/// We elect to use an ordering that minimizes space when stored in a `table::Page`, since this
+/// means we don't have to re-order the data when we read it in / back out.
+pub fn alignment(type_: &AlgebraicType, ctx: &Typespace) -> u8 {
+    match type_ {
+        &AlgebraicType::Bool | &AlgebraicType::I8 | &AlgebraicType::U8 => 1,
+        &AlgebraicType::I16 | &AlgebraicType::U16 => 2,
+        &AlgebraicType::I32 | &AlgebraicType::U32 | &AlgebraicType::F32 => 4,
+        &AlgebraicType::I64 | &AlgebraicType::U64 | &AlgebraicType::F64 => 8,
+        &AlgebraicType::I128 | &AlgebraicType::U128 => 16,
+        // The alignment of the following cases being 2 may seem slightly surprising.
+        // This is based on the alignment when stored in a page in the `table` crate.
+        &AlgebraicType::String
+        | AlgebraicType::Builtin(BuiltinType::Array(_))
+        | AlgebraicType::Builtin(BuiltinType::Map(_)) => 2,
+        // Minimum possible alignment is 1, even though minimum possible size is 0.
+        // This is consistent with Rust.
+        AlgebraicType::Product(ref product) => product
+            .elements
+            .iter()
+            .map(|child| alignment(&child.algebraic_type, ctx))
+            .max()
+            .unwrap_or(1),
+        AlgebraicType::Sum(ref sum) => sum
+            .variants
+            .iter()
+            .map(|variant| alignment(&variant.algebraic_type, ctx))
+            .max()
+            .unwrap_or(1),
+        AlgebraicType::Ref(ref ref_) => alignment(&ctx[*ref_], ctx),
+    }
+}
```

Comment:
I do not feel good about duplicating alignment computation in addition to the version in the table crate (in `layout.rs`). It may be unavoidable, but if so please add comments *to both implementations* to the effect that:
1. It is very important that they stay consistent.
2. Why the duplication is necessary.

Reply:
Removed!